src/features/profile/profileSlice.ts

```diff
@@ -0,0 +1,161 @@
+  user: KBaseProfileUser;
+  profile: Meta;
+}
+
+interface KBaseProfileResponseCache extends Meta {
+  [user: string]: KBaseProfileResponse;
+}
+
+// Define a type for the slice state
+export interface ProfileState extends Meta {
+  cache: KBaseProfileResponseCache;
+  error: string;
+  pending: boolean;
+  realname: string;
+  usernameSelected: string | null;
+}
+
+const initialState: ProfileState = {
+  cache: {},
+  error: '',
+  pending: false,
+  realname: '',
+  usernameSelected: null,
+};
+
+const nullProfileFactory = (username: string): KBaseProfileResponseCache => ({
+  profile: {
+    user: {
+      username,
+      realname: '',
+    },
+    profile: {},
+  },
+});
+
+interface LoadProfileParams {
+  token: string;
+  username?: string;
+}
+
+export const loadProfiles = async (token: string, usernames: string[]) => {
+  const client = new KBaseServiceClient({
+    module: 'UserProfile',
+    url: 'https://ci.kbase.us/services/user_profile/rpc',
+    authToken: token,
+  });
+  const profiles = await client.call('get_user_profile', [usernames]);
+  return profiles;
+};
+
+export const cache = ({ profile }: { profile: ProfileState }) => profile.cache;
+export const loadProfile = createAsyncThunk(
+  'profile/loadProfile',
+  async (
+    { token, username }: LoadProfileParams,
+    { getState, rejectWithValue }
+  ): Promise<KBaseProfileResponseCache> => {
+    if (!username) throw new Error(errMsgUsername);
+    const profilesCached = cache(getState() as { profile: ProfileState });
+    if (username in profilesCached) {
+      return { profile: profilesCached[username] };
+    }
+    let profiles;
+    try {
+      profiles = await loadProfiles(token, [username]);
+    } catch (err) {
+      if (err instanceof KBaseJsonRpcError) {
+        throw new Error(errMsgKBase);
+      }
+      throw err;
+    }
+    const profile: KBaseProfileResponse = profiles[0];
+    if (!profile) {
+      return Promise.reject(
+        rejectWithValue({
+          username,
+          error: `Username '${username}' does not exist.`,
+        })
+      );
+    }
+    return { profile };
+  }
+);
+
+export const profileSlice = createSlice({
+  name: 'profile',
+  initialState,
+  reducers: {
+    clearError: (state) => {
+      state.error = '';
+    },
+    selectUsername: (state, action) => {
+      state.usernameSelected = action.payload;
+    },
+  },
+  extraReducers: (builder): ActionReducerMapBuilder<ProfileState> =>
+    builder
+      .addCase(loadProfile.pending, (state) => {
+        state.error = '';
+        state.pending = true;
+      })
+      .addCase(loadProfile.fulfilled, (state, action) => {
+        const { profile: profileCache } =
+          action.payload as KBaseProfileResponseCache;
+        const { user } = profileCache;
+        state.realname = user.realname;
+        state.usernameSelected = user.username;
+        const cache: KBaseProfileResponseCache = {
+          [state.usernameSelected]: profileCache,
+        };
```

Comment:
Wouldn't this only allow one profile to be cached at a time? I'm assuming you want to cache more than one. Wouldn't you want the logic here to be something more like this?:
```typescript
const cache: KBaseProfileResponseCache = {
  [state.usernameSelected]: profileCache,
  ...state.cache
}
```

Reply:
Its possible that you'd only want to cache one at a time, but if that's true then I'm not sure why `cache` would need to be a key value store

Reply:
Fixed in the latest commit.